### PR TITLE
Add support for VkMemoryBarrier2KHR and VkPipelineDiscardRectangleStateCreateInfoEXT

### PR DESCRIPTION
**Pull request by hakzsam**
This also includes a fix for missing hashing of graphics/compute pipelines pNext.

### CODE REVIEW

File: fossilize.cpp [hunk 1]
```diff
@@ -3752,6 +3775,24 @@ bool StateReplayer::Impl::parse_input_attachment_aspect(const Value &state,
...
 
+bool StateReplayer::Impl::parse_discard_rectangles(const Value &state,
+						   VkPipelineDiscardRectangleStateCreateInfoEXT **out_info)
+{
+	auto *info = allocator.allocate_cleared<VkPipelineDiscardRectangleStateCreateInfoEXT>();
+	*out_info = info;
+
+	info->flags = static_cast<VkPipelineDiscardRectangleStateCreateFlagsEXT>(state["flags"].GetUint());
+	info->discardRectangleMode = static_cast<VkDiscardRectangleModeEXT>(state["discardRectangleMode"].GetUint());
+	info->discardRectangleCount = state["discardRectangleCount"].GetUint();
```
Comment (HansKristian-Work) [inline fossilize.cpp:3809]: Don't store separate count variable.

File: fossilize.cpp [hunk 2]
```diff
@@ -4489,6 +4539,15 @@ void *StateRecorder::Impl::copy_pnext_struct(const VkRenderPassInputAttachmentAs
 	return input_att;
 }
 
+void *StateRecorder::Impl::copy_pnext_struct(const VkPipelineDiscardRectangleStateCreateInfoEXT *create_info,
+                                             ScratchAllocator &alloc)
+{
+	auto *discard_rectangles = copy(create_info, 1, alloc);
+	if (create_info->pDiscardRectangles)
```
Comment (HansKristian-Work) [inline fossilize.cpp:4596]: Better to check discard_rectangles->discardRectangleCount here I think.

File: fossilize.cpp [hunk 3]
```diff
@@ -6927,6 +6993,35 @@ static bool json_value(const VkRenderPassInputAttachmentAspectCreateInfo &create
...
 }
 
+template <typename Allocator>
+static bool json_value(const VkPipelineDiscardRectangleStateCreateInfoEXT &create_info, Allocator &alloc, Value *out_value)
+{
+	Value value(kObjectType);
+	value.AddMember("sType", create_info.sType, alloc);
+	value.AddMember("flags", create_info.flags, alloc);
+	value.AddMember("discardRectangleMode", create_info.discardRectangleMode, alloc);
+	value.AddMember("discardRectangleCount", create_info.discardRectangleCount, alloc);
```
Comment (HansKristian-Work) [inline fossilize.cpp:7074]: Or hm ... Perhaps we need separate count variable if discard rectangle is considered dynamic?

Comment (HansKristian-Work) [reply fossilize.cpp:7074]: Yes, we store viewportCount/scissorCount for this reason, so ...

**Comment by hakzsam** [comment]
Rebased on top of the new dynamic state work.

File: fossilize.cpp [hunk 4]
```diff
@@ -957,6 +965,41 @@ static void hash_pnext_struct(const StateRecorder *,
...
+static bool hash_pnext_struct(const StateRecorder *recorder,
+                              Hasher &h,
+                              const VkMemoryBarrier2KHR &info)
+{
+	h.u32(info.srcStageMask);
+	h.u32(info.srcAccessMask);
+	h.u32(info.dstStageMask);
+	h.u32(info.dstAccessMask);
+
+	if (!hash_pnext_chain(recorder, h, info.pNext, nullptr))
```
Comment (HansKristian-Work) [inline fossilize.cpp:997]: Don't hash this here. It's part of the pNext chain and will be picked up.

File: fossilize.cpp [hunk 5]
```diff
@@ -3758,6 +3819,44 @@ bool StateReplayer::Impl::parse_input_attachment_aspect(const Value &state,
...
+{
+	auto *info = allocator.allocate_cleared<VkMemoryBarrier2KHR>();
+	*out_info = info;
+
+	info->srcStageMask = static_cast<VkPipelineStageFlags2KHR>(state["srcStageMask"].GetUint());
+	info->srcAccessMask = static_cast<VkAccessFlags2KHR>(state["srcAccessMask"].GetUint());
+	info->dstStageMask = static_cast<VkPipelineStageFlags2KHR>(state["dstStageMask"].GetUint());
+	info->dstAccessMask = static_cast<VkAccessFlags2KHR>(state["dstAccessMask"].GetUint());
+
+	if (state.HasMember("pNext"))
```
Comment (HansKristian-Work) [inline fossilize.cpp:3853]: This pnext is part of a chain and shouldn't be copied here.

File: fossilize.cpp [hunk 6]
```diff
@@ -4499,6 +4606,29 @@ void *StateRecorder::Impl::copy_pnext_struct(const VkRenderPassInputAttachmentAs
...
+		discard_rectangles->pDiscardRectangles = nullptr;
+	return discard_rectangles;
+}
+
+void *StateRecorder::Impl::copy_pnext_struct(const VkMemoryBarrier2KHR *create_info,
+                                             ScratchAllocator &alloc)
+{
+	auto *memory_barrier2 = copy(create_info, 1, alloc);
+
+	if (!copy_pnext_chain(memory_barrier2->pNext, alloc, &memory_barrier2->pNext, nullptr))
```
Comment (HansKristian-Work) [inline fossilize.cpp:4626]: Stray copy that shouldn't be there.

File: fossilize.cpp [hunk 7]
```diff
@@ -4674,6 +4804,23 @@ bool StateRecorder::Impl::copy_pnext_chain(const void *pNext, ScratchAllocator &
...
+			auto *ci = static_cast<const VkPipelineDiscardRectangleStateCreateInfoEXT *>(pNext);
+			*ppNext = static_cast<VkBaseInStructure *>(copy_pnext_struct(ci, alloc, dynamic_state_info));
+			break;
+		}
+
+		case VK_STRUCTURE_TYPE_MEMORY_BARRIER_2_KHR:
+		{
+			auto *ci = static_cast<const VkMemoryBarrier2KHR *>(pNext);
+			auto *memory_barrier2 = static_cast<VkBaseInStructure *>(copy_pnext_struct(ci, alloc));
+			if (!memory_barrier2)
```
Comment (HansKristian-Work) [inline fossilize.cpp:4818]: Redundant.

File: fossilize.cpp (hunk 6 above)
Comment (hakzsam) [reply fossilize.cpp:4626]: Removed.

File: fossilize.cpp (hunk 7 above)
Comment (hakzsam) [reply fossilize.cpp:4818]: Removed.

**Pull request merged**